File: decompiled/c/411b58.c

undefined4 FUN_00411b58(int *param_1,LPCSTR param_2)

{
  undefined1 uVar1;
  HGLOBAL hResData;
  int local_458 [3];
  ushort local_44a;
  uint local_438;
  undefined1 local_430 [16];
  uint local_420;
  undefined1 local_41c [1024];
  uint *local_1c;
  uint local_18;
  undefined4 local_14;
  HRSRC local_10;
  HFILE local_c;
  int local_8;
  
  for (local_18 = 0; (int)local_18 < 0x100; local_18 = local_18 + 1) {
    local_41c[local_18 * 4] = (char)(((local_18 >> 5 & 7) * 0xff) / 7);
    local_41c[local_18 * 4 + 1] = (char)(((local_18 >> 2 & 7) * 0xff) / 7);
    local_41c[local_18 * 4 + 2] = (char)(((local_18 & 3) * 0xff) / 3);
    local_41c[local_18 * 4 + 3] = 0;
  }
  if ((param_2 == (LPCSTR)0x0) ||
     (local_10 = FindResourceA((HMODULE)0x0,param_2,(LPCSTR)0x2), local_10 == (HRSRC)0x0)) {
    if ((param_2 != (LPCSTR)0x0) && (local_c = _lopen(param_2,0), local_c != -1)) {
      _lread(local_c,local_430,0xe);
      _lread(local_c,local_458,0x28);
      _lread(local_c,local_41c,0x400);
      _lclose(local_c);
      if (local_458[0] == 0x28) {
        if (local_44a < 9) {
          if (local_438 == 0) {
            local_420 = 1 << ((byte)local_44a & 0x1f);
          }
          else {
            local_420 = local_438;
          }
        }
        else {
          local_420 = 0;
        }
      }
      else {
        local_420 = 0;
      }
      for (local_18 = 0; (int)local_18 < (int)local_420; local_18 = local_18 + 1) {
        uVar1 = local_41c[local_18 * 4];
        local_41c[local_18 * 4] = local_41c[local_18 * 4 + 2];
        local_41c[local_18 * 4 + 2] = uVar1;
      }
    }
  }
  else {
    hResData = LoadResource((HMODULE)0x0,local_10);
    local_1c = LockResource(hResData);
    local_8 = *local_1c + (int)local_1c;
    if ((local_1c == (uint *)0x0) || (*local_1c < 0x28)) {
      local_420 = 0;
    }
    else if (*(ushort *)((int)local_1c + 0xe) < 9) {
      if (local_1c[8] == 0) {
        local_420 = 1 << (*(byte *)((int)local_1c + 0xe) & 0x1f);
      }
      else {
        local_420 = local_1c[8];
      }
    }
    else {
      local_420 = 0;
    }
    for (local_18 = 0; (int)local_18 < (int)local_420; local_18 = local_18 + 1) {
      local_41c[local_18 * 4] = *(undefined1 *)(local_8 + 2 + local_18 * 4);
      local_41c[local_18 * 4 + 1] = *(undefined1 *)(local_8 + 1 + local_18 * 4);
      local_41c[local_18 * 4 + 2] = *(undefined1 *)(local_8 + local_18 * 4);
      local_41c[local_18 * 4 + 3] = 0;
    }
  }
  (**(code **)(*param_1 + 0x14))(param_1,4,local_41c,&local_14,0);
  return local_14;
}

File: ported/auto/411b58.js
// @manual — do not regenerate.
// Source: decompiled/c/411b58.c
// Fix: local_41c is a byte[1024] (RGBA pixel buffer); translator emitted
// setU32 for byte stores and doubled the index (`(i*4)*4`). Fixed three
// loops (gradient init, BGR swap, BMP-load) — all setU32→setU8,
// heap.u32→heap.u8 on reads, removed extra `*4`.

/** @typedef {import("../../runtime/heap.js").Heap} Heap */

import { FindResourceA, LoadResource, LockResource, _lclose, _lopen, _lread } from "../../runtime/win32.js";
import { callIndirect } from "../../runtime/win32/context.js";
import { regs } from "../../runtime/regs.js";
export function FUN_00411b58(heap, param_1, param_2) {
  const __sp = heap.allocFrame(1112);
  const __addr_local_14 = __sp + 1096;
  const __addr_local_458 = __sp + 4;
  const __addr_local_430 = __sp + 44;
  const __addr_local_41c = __sp + 64;
  const __addr_local_44a = __sp + 18;
  const __addr_local_438 = __sp + 36;
  const __addr_local_420 = __sp + 60;
  const __addr_local_1c = __sp + 1088;
  const __addr_local_18 = __sp + 1092;
  const __addr_local_10 = __sp + 1100;
  const __addr_local_c = __sp + 1104;
  const __addr_local_8 = __sp + 1108;
  try {
  let uVar1 = 0;
  let hResData = 0;
  for (heap.setU32(__addr_local_18, (0) >>> 0); ((heap.u32(__addr_local_18)) | 0) < 0x100; heap.setU32(__addr_local_18, (heap.u32(__addr_local_18) + 1) >>> 0)) {
    heap.setU8((__addr_local_41c + heap.u32(__addr_local_18) * 4), ((((((heap.u32(__addr_local_18) >>> 5 & 7) * 0xff) / 7)) << 24 >> 24)) & 0xff);
    heap.setU8((__addr_local_41c + heap.u32(__addr_local_18) * 4 + 1), ((((((heap.u32(__addr_local_18) >>> 2 & 7) * 0xff) / 7)) << 24 >> 24)) & 0xff);
    heap.setU8((__addr_local_41c + heap.u32(__addr_local_18) * 4 + 2), ((((((heap.u32(__addr_local_18) & 3) * 0xff) / 3)) << 24 >> 24)) & 0xff);
    heap.setU8((__addr_local_41c + heap.u32(__addr_local_18) * 4 + 3), (0) & 0xff);
  }
  if ((param_2 == ((0x0) | 0)) || (heap.setU32(__addr_local_10, (FindResourceA(heap, ((0x0) | 0), param_2, ((0x2) | 0))) >>> 0), heap.u32(__addr_local_10) == ((0x0) | 0))) {
    if ((param_2 != ((0x0) | 0)) && (heap.setU32(__addr_local_c, (_lopen(heap, param_2, 0)) >>> 0), (heap.u32(__addr_local_c) | 0) != -1)) {
      _lread(heap, heap.u32(__addr_local_c), __addr_local_430, 0xe);
      _lread(heap, heap.u32(__addr_local_c), __addr_local_458, 0x28);
      _lread(heap, heap.u32(__addr_local_c), __addr_local_41c, 0x400);
      _lclose(heap, heap.u32(__addr_local_c));
      if (heap.u32(__addr_local_458 + (0) * 4) == 0x28) {
        if (heap.u32(__addr_local_44a) < 9) {
          if (heap.u32(__addr_local_438) == 0) {
            heap.setU32(__addr_local_420, (1 << (((heap.u32(__addr_local_44a)) & 0xff) & 0x1f)) >>> 0);
          } else {
            heap.setU32(__addr_local_420, (heap.u32(__addr_local_438)) >>> 0);
          }
        } else {
          heap.setU32(__addr_local_420, (0) >>> 0);
        }
      } else {
        heap.setU32(__addr_local_420, (0) >>> 0);
      }
      for (heap.setU32(__addr_local_18, (0) >>> 0); ((heap.u32(__addr_local_18)) | 0) < ((heap.u32(__addr_local_420)) | 0); heap.setU32(__addr_local_18, (heap.u32(__addr_local_18) + 1) >>> 0)) {
        uVar1 = ((heap.u8(__addr_local_41c + heap.u32(__addr_local_18) * 4)) & 0xff);
        heap.setU8((__addr_local_41c + heap.u32(__addr_local_18) * 4), (heap.u8(__addr_local_41c + heap.u32(__addr_local_18) * 4 + 2)) & 0xff);
        heap.setU8((__addr_local_41c + heap.u32(__addr_local_18) * 4 + 2), (uVar1) & 0xff);
      }
    }
  } else {
    hResData = ((LoadResource(heap, ((0x0) | 0), heap.u32(__addr_local_10))) >>> 0);
    heap.setU32(__addr_local_1c, (LockResource(heap, hResData)) >>> 0);
    heap.setU32(__addr_local_8, (heap.u32(heap.u32(__addr_local_1c)) + ((heap.u32(__addr_local_1c)) | 0)) >>> 0);
    if ((heap.u32(__addr_local_1c) == 0x0) || (heap.u32(heap.u32(__addr_local_1c)) < 0x28)) {
      heap.setU32(__addr_local_420, (0) >>> 0);
    } else {
      if (heap.u16((((heap.u32(__addr_local_1c)) | 0) + 0xe)) < 9) {
      if (heap.u32(heap.u32(__addr_local_1c) + (8) * 4) == 0) {
        heap.setU32(__addr_local_420, (1 << (heap.u8((((heap.u32(__addr_local_1c)) | 0) + 0xe)) & 0x1f)) >>> 0);
      } else {
        heap.setU32(__addr_local_420, (heap.u32(heap.u32(__addr_local_1c) + (8) * 4)) >>> 0);
      }
    } else {
      heap.setU32(__addr_local_420, (0) >>> 0);
    }
    }
    for (heap.setU32(__addr_local_18, (0) >>> 0); ((heap.u32(__addr_local_18)) | 0) < ((heap.u32(__addr_local_420)) | 0); heap.setU32(__addr_local_18, (heap.u32(__addr_local_18) + 1) >>> 0)) {
      heap.setU8((__addr_local_41c + heap.u32(__addr_local_18) * 4), (heap.u8((heap.u32(__addr_local_8) + 2 + heap.u32(__addr_local_18) * 4))) & 0xff);
      heap.setU8((__addr_local_41c + heap.u32(__addr_local_18) * 4 + 1), (heap.u8((heap.u32(__addr_local_8) + 1 + heap.u32(__addr_local_18) * 4))) & 0xff);
      heap.setU8((__addr_local_41c + heap.u32(__addr_local_18) * 4 + 2), (heap.u8((heap.u32(__addr_local_8) + heap.u32(__addr_local_18) * 4))) & 0xff);
      heap.setU8((__addr_local_41c + heap.u32(__addr_local_18) * 4 + 3), (0) & 0xff);
    }
  }
  (regs.eax = callIndirect(heap, heap.u32((heap.i32(param_1) + 0x14)), param_1, 4, __addr_local_41c, __addr_local_14, 0));
  return heap.u32(__addr_local_14);
} finally {
    heap.freeFrame(1112);
  }
}
